### PR TITLE
[WIP] Entry theme selection

### PR DESCRIPTION
**Pull request by aviav**

### CODE REVIEW

File: app/assets/javascripts/pageflow/editor/views/theme_view.js [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+    },
+    'mouseleave': function() {
+      var that = this;
+      this.uiPreviewImage.attr('src', pageflow.editor.themes.find(function(theme) {
+        return theme.attributes.name == pageflow.entry.configuration.attributes.theme_name;
+      }).get('preview_image_url'));
+      this.uiPreviewHeaderThemeName.text(that.translateThemeName({
+        name: pageflow.entry.configuration.attributes.theme_name
+      }));
+    }
```
Comment (tf) [inline app/assets/javascripts/pageflow/editor/views/theme_view.js]: Depending on DOM-elements of other views is an anti-pattern. This logic already exists in `ChangeThemeDialogView` and is duplicated here. Instead, `ChangeThemeDialogView` and `ThemeView` could share a `selection` Backbone.Model which is updated by the `ThemeView` and used by the dialog view to decide which preview to display. See [the `FilesExplorerView`](https://github.com/codevise/pageflow/blob/master/app/assets/javascripts/pageflow/editor/views/files_explorer_view.js#L24) for an example.

File: app/assets/javascripts/pageflow/editor/views/theme_view.js [hunk 2]
```diff
@@ -0,0 +1,67 @@
+/** @api private */
+pageflow.ThemeView = Backbone.Marionette.ItemView.extend({
```
Comment (tf) [inline app/assets/javascripts/pageflow/editor/views/theme_view.js]: Most of the views which are supposed to be item views in a collection view all end with the suffix `ItemView`.

File: app/assets/stylesheets/pageflow/editor/dialogs.scss [hunk 3]
```diff
@@ -5,6 +5,7 @@
   width: 100%;
   height: 100%;
   z-index: 5;
+  background-color: transparent;
```
Comment (tf) [inline app/assets/stylesheets/pageflow/editor/dialogs.scss]: Why is this needed?

File: app/assets/javascripts/pageflow/editor/views/theme_view.js (hunk 1 above)
Comment (aviav) [inline app/assets/javascripts/pageflow/editor/views/theme_view.js]: I agree. Adding to this line of thought, I think it is more appropriate in this case to even use a `preview` Backbone.Model in addition to a `selection` Backbone.Model, since previewed theme and selected theme can differ. This implies reusing `selectableView` and writing a very similar `previewableView`.

Comment (tf) [inline app/assets/javascripts/pageflow/editor/views/theme_view.js]: I think the current handling of the active theme (the one that is used in the entry) is fine. The view never has to react to a change, but only has to find out if it is active initially. While I agree that `preview` might be a better name, I would also be totally fine with calling it `selection` if that makes reusing `selectableView` easier.

Comment (aviav) [inline app/assets/javascripts/pageflow/editor/views/theme_view.js]: I think this would make our code unnecessarily hard to read. We have a selection workflow and a previewing workflow accessible from here, and we would reuse a selection mixin for the preview workflow while the selection workflow uses its own code but does something which is similar to what the selection mixin does. As soon as somebody else works on this code, there is significant risk that they are confused by this.

I will see if we can work with two mixins here without much additional effort

File: app/assets/stylesheets/pageflow/editor/dialogs.scss (hunk 3 above)
Comment (aviav) [inline app/assets/stylesheets/pageflow/editor/dialogs.scss]: It is just an artifact of trying to reproduce exactly the styles of the text tracks dialog. Testing revealed that it is not needed. I just thought it was

Comment (aviav) [inline app/assets/stylesheets/pageflow/editor/dialogs.scss]: `background: transparent`, which we already have on `master`, is needed though

File: app/assets/javascripts/pageflow/editor/views/change_theme_view.js [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+      });
+    }
+  },
+
+  onRender: function() {
+    this.ui.labelText.text(this.labelText());
+  },
+
+  labelText: function() {
+    return this.options.label || this.labelPrefix() + ': ' + this.localizedThemeName();
```
Comment (aviav) [inline app/assets/javascripts/pageflow/editor/views/change_theme_view.js]: Do you have anything specific in mind here? The most convenient and visually consistent kind of element I can think of is a select input as a way to quickly switch themes. We could disable this when no more than one theme is available

Comment (tf) [inline app/assets/javascripts/pageflow/editor/views/change_theme_view.js]: That would be a duplication since he theme selection dialog already takes the role of that select input. I'd just put the name of the current theme below the label as plain text, much like the name of the current file is currently displayed in file input views. Then the user can click the button to open the dialog and make changes.

File: app/assets/javascripts/pageflow/editor/views/theme_item_view.js [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+  events: {
+    'click .use_theme': function(event) {
+      this.configuration.set('theme_name', this.model.get('name'));
+      pageflow.updateStylesheet('theme');
+    },
+    'mouseenter': function() {
+      this.select();
+    },
+    'mouseleave': function() {
+      this.onClose();
```
Comment (tf) [inline app/assets/javascripts/pageflow/editor/views/theme_item_view.js]: Calling the `onClose` lifecycle handler here feels a bit unidiomatic. The fact that `selectableView` implements `onClose` is more of an implementation detail. Even more, I am wondering if this is necessary at all. I think it would be just fine if the most recently hovered theme stayed selected even after the mouse left the item. This would also be more in line with a touch friendly interface: We could, for example, bind both `mouseenter` and `click` to ensure themes can be selected on platforms without mouse events.

The only thing we would need to add is some visual hint (like a changed background color) indicating which theme is currently selected.

File: app/assets/javascripts/pageflow/editor/utils/update_stylesheet.js [hunk 6]
```diff
@@ -0,0 +1,11 @@
...
+  var link = $('head link[data-name=' + name + ']');
+
+  var newHref = pageflow.editor.themes.find(function(theme) {
+    return theme.attributes.name == pageflow.entry.configuration.attributes.theme_name;
+  }).get('stylesheet_path');
+
+  if (link.attr('href') !== newHref) {
+    link.attr('href', newHref);
+  }
+};
```
Comment (tf) [inline app/assets/javascripts/pageflow/editor/utils/update_stylesheet.js]: Same goes for the logic that finds a theme by name, which is also repeated in a couple of places.

File: app/assets/javascripts/pageflow/editor/initializers/setup_themes.js [hunk 7]
```diff
@@ -0,0 +1,3 @@
+pageflow.app.addInitializer(function(options) {
+  pageflow.editor.themes = new Backbone.Collection(options.themes);
```
Comment (tf) [inline app/assets/javascripts/pageflow/editor/initializers/setup_themes.js]: Same goes for the logic that finds a theme by name, which is also repeated in a couple of places.

File: app/assets/javascripts/pageflow/editor/views/change_theme_view.js (hunk 4 above)
Comment (aviav) [inline app/assets/javascripts/pageflow/editor/views/change_theme_view.js]: Duplication from a code/data perspective only. It makes a big difference for some users whether they can change something from a dropdown without going another layer of nesting deeper (feature 1), and it makes a big difference for some users whether they can preview graphical features of their themes in one place (feature 2).

Be that as it may, I shall go with your recommendation because it is simpler

Comment (aviav) [inline app/assets/javascripts/pageflow/editor/views/change_theme_view.js]: Scenario for feature 1: I have a customer presentation with a very limited timeframe and want to quickly present the customer with several choices of themes. The customer will be able to see consequences of their choice directly in the editor, together with their media files.

Alternative, without feature 1: The presenter will have to open the theme picker dialog several times or has to resort to a preview that uses static image files, which are in some ways less close to the final product and create needless room for interpretation/questions that way

Comment (aviav) [inline app/assets/javascripts/pageflow/editor/views/change_theme_view.js]: Of course it would be nice to generate the preview images using actual media files from the entry, but that would tax computer and developer significantly

Comment (tf) [inline app/assets/javascripts/pageflow/editor/views/change_theme_view.js]: I think, all of this is feature creep. Let's just go with the plain and simple text for now.

**Comment by aviav** [comment]
rebased onto `master`
